### PR TITLE
Proposal about Jupyter computing, Scientifik plot, Scientifik DSL

### PR DESCRIPTION
**Pull request by Zelenyy**

### CODE REVIEW

**Comment by altavir** [comment]
Please add an expanded description of the proposal in the pull request.

File: use-cases/scientific-plot.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Scientifik chart, plot and etc
```
Comment (altavir) [inline use-cases/scientific-plot.md:1]: Is not relevant for this repository. Follow discussion in https://kotlinlang.slack.com/messages/CEXV2QWNM/, there will be additional repository for visualization proposal. Visualization is important but should be decoupled from mathematical API.

File: use-cases/notebooks-computation.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+
+# Jupyter notebook
```
Comment (altavir) [inline use-cases/notebooks-computation.md:2]: See comment below about visualization. Currently jupyter supports a few different kotlin plugin. We will probably want for beakerx to support our maths in their autotranslation feature.

The statement is that you need imperative statements for notebook is completely wrong. See plotly for example.

File: use-cases/fast-estimation.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```python
+np.corrcoef(data['a1'], data['a2'])
+```
+    array([[ 1.        ,  0.41185499],
+           [ 0.41185499,  1.        ]])
+
+Several line of code  and I already known, what my supervisor will send me to remake all work.
+
+How conclusions can we make from this example? 
+Какие полезные выводы мы можем сделать из этого примера? It can be noted that this code is declarative rather than imperative, I do not describe in detail what I should do with the data, I just say: load data, make histogramm. For comparasion, for loading data on Kotlin, I must make this for example:
```
Comment (altavir) [inline use-cases/fast-estimation.md]: Some Russian left here.

Comment (altavir) [inline use-cases/fast-estimation.md]: > this code is declarative rather than imperative

I do not agree with that statement. Declarative means what you want to achieve, not what you actually do to achieve that, but it is question of definitions.

**Comment by Zelenyy** [comment]
Proposal description:

- fast-estimation.md - proposal for develop user-friendly DSL kind of like Matplab/Python.
- notebooks-computation.md - proposal for integration with Jupyter
- scientific-plot.md - description of plot use case in science.

File: use-cases/fast-estimation.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            } while (num > 0)
+        }
+    }
+}
+```
+Only data loading request code more than all processing on Python. Of course in a big project this will not be a problem, but in time when I want just  fast to look at data for checking  some simple ideas, I don't want to write a sheet of code. I want to make something as:
+```kotlin
+val data = read(filename, type=Data:class)
+```
+Or building of histogramm: for example, if I want use  [hep.aida](https://dst.lbl.gov/ACSSoftware/colt/api/hep/aida/package-summary.html), I must be create factory, next create histogramm and only later I can feel histogramm --- three action instead one action.
```
Comment (altavir) [inline use-cases/fast-estimation.md:76]: AIDA factory is a low level API in kotlin you can safely hide everything under the hood. It is in fact much better way than the python global state changes.

File: use-cases/fast-estimation.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+}
+```
+Only data loading request code more than all processing on Python. Of course in a big project this will not be a problem, but in time when I want just  fast to look at data for checking  some simple ideas, I don't want to write a sheet of code. I want to make something as:
+```kotlin
+val data = read(filename, type=Data:class)
+```
+Or building of histogramm: for example, if I want use  [hep.aida](https://dst.lbl.gov/ACSSoftware/colt/api/hep/aida/package-summary.html), I must be create factory, next create histogramm and only later I can feel histogramm --- three action instead one action. 
+Или же построение гистограммы: например при использовании, исследователь должен создать фабрику, затем создать гистограмму, и только потом он может гистограмировать, три действия там где достаточно одного.In addition, there is a separate issue of performance, for example, the existing solution in [kmath](https://github.com/altavir/kmath) is good in terms of thread safety, but memory consumption in the case of processing data with large dimensions is redundant and leads to high GC load.
```
Comment (altavir) [inline use-cases/fast-estimation.md]: Another Russian line. I can read it. Others don't

**Review by altavir — CHANGES_REQUESTED** [review]
The general idea of proposed use-cases is clear. People need something quick and dirty to work like in python. The description is not very clear and concise though and won't be accepted as is for several reasons:

1. We wont it work like Python is a bad argument because in most cases people are coming to kotlin because they are not satisfied with Python. Python does a lot of things in a wrong way.

2. The example is very limited. Obviously you are talking about multi-dimensional histograms, but you should state what you want to do with them. You want fixed bin size, or custom bin size, you want to optimize performance on read or on write, do you want to implement an algebra over histograms or you want to treat them as nd-structures?

3. Visualization and IO are separate from mathematic API. As I have written in comments, there will be additional repository for plots. As for output. Probably we need to add a request into feature section about it, but it requires a lot of thinking.

In conclusion, this PR will be accepted only after removing all parts not connected to histograms, change of file name and more thorough explanation of requirements.

File: use-cases/scientific-plot.md (hunk 1 above)
Comment (Zelenyy) [reply use-cases/scientific-plot.md:1]: I will give a few objections.

Firstly, on python people can code quick and dirty, and Our aim is possibility to code quick and clear. If its willn't quick, that its is dead. Hascell is very clear and very unpopular.

Secondly, wrong to think, what people are not satisfied with Python. We must concurrent with Python and figth to weak side of Python, Numpy API isn't weak side. Weak side of Python is perfomance, GIL, darkwing duck typing (where in really I very rather use dynamic typing). 

Thirdly,
> You want fixed bin size, or custom bin size, you want to optimize performance on read or on write, do you want to implement an algebra over histograms or you want to treat them as nd-structures?

The whole point is that I ( as library user) don't want to think about it, and our target save user from this.

I agree what IO and plot tools must be independ API, but analysis of proposal should make for whole ecosystem of Scientific library.

File: use-cases/notebooks-computation.md (hunk 2 above)
Comment (Zelenyy) [reply use-cases/notebooks-computation.md:2]: I'm probably  wrong to write, I don't request imperative statements for jupyter.

File: use-cases/fast-estimation.md (hunk 4 above)
Comment (Zelenyy) [reply use-cases/fast-estimation.md:76]: `histogramm` in Python is function, there is'not any state))

>AIDA factory is a low level API in kotlin you can safely hide everything under the hood.

This is my main aim.

File: use-cases/fast-estimation.md (hunk 5 above)
Comment (Zelenyy) [reply use-cases/fast-estimation.md]: yes, I was distract and forget delete russinas line.

File: use-cases/fast-estimation.md (hunk 3 above)
Comment (Zelenyy) [reply use-cases/fast-estimation.md]: Yes, and in Numpy, I say: "I want histogramm", and Numpy make histogram, I don't think detail, I just get histogramm and see on its.

File: use-cases/scientific-plot.md (hunk 1 above)
Comment (altavir) [reply use-cases/scientific-plot.md:1]: For that you usually have some kind of top-level sugar. We can't start development based on that. Even so, to guess what user will want, we need to understand what to do. So use-case must have that information. Otherwise it is useless.
